tests/test_avro_schema.py

```diff
@@ -8,6 +8,206 @@
 import json
 import pytest
 
+# Schemas defined in AvroCompatibilityTest.java. Used here to ensure compatibility with the schema-registry
+schema1 = parse('{"type":"record","name":"myrecord","fields":[{"type":"string","name":"f1"}]}')
+schema2 = parse(
+    '{"type":"record","name":"myrecord","fields":[{"type":"string","name":"f1"},{"type":"string",'
+    '"name":"f2","default":"foo"}]}'
+)
+schema3 = parse('{"type":"record","name":"myrecord","fields":[{"type":"string","name":"f1"},{"type":"string","name":"f2"}]}')
+schema4 = parse('{"type":"record","name":"myrecord","fields":[{"type":"string","name":"f1_new","aliases":["f1"]}]}')
+schema6 = parse('{"type":"record","name":"myrecord","fields":[{"type":["null","string"],"name":"f1","doc":"doc of f1"}]}')
+schema7 = parse(
+    '{"type":"record","name":"myrecord","fields":[{"type":["null","string","int"],"name":"f1","doc":"doc of f1"}]}'
+)
+# schema8 = (
+#     '{"type":"record","name":"myrecord","fields":[{"type":"string","name":"f1"},{"type":"string",'
+#     '"name":"f2","default":"foo"}]},{"type":"string","name":"f3","default":"bar"}]}'
+# )
```

Comment:
this JSON is invalid, which is probably causing the tests related to it to fail. As a rule of thumb, i find that `parse(dumps..` is safer than `parse`

Reply:
> this JSON is invalid, which is probably causing the tests related to it to fail. 

Exactly. I can't call `parse` here because an exception is raised.

The problem is that this is how the tests are originally defined [ref](https://github.com/confluentinc/schema-registry/blob/e8f89f8e55ac3f2a3c32abbd9f4de0affff238c8/core/src/test/java/io/confluent/kafka/schemaregistry/avro/AvroCompatibilityTest.java#L76). I commented it out until I have time to dig into the java implementation and see how it is handling the invalid JSON.

Reply:
Okay, so it seems the tests don't care if there is trailing data when parsing. I added a test to print the contents of the `schema8` and this is what I got:

```java
  @Test
  public void testPrintSchemaEight() {
    System.out.println(schema8);
  }
```
> {"type":"record","name":"myrecord","fields":[{"name":"f1","type":"string"},{"name":"f2","type":"string","default":"foo"}]}

Now I have to check if this is also done for production code.

Reply:
Okay, so:

- [The tests uses `AvroUtils.parseSchema`](https://github.com/confluentinc/schema-registry/blob/e8f89f8e55ac3f2a3c32abbd9f4de0affff238c8/core/src/test/java/io/confluent/kafka/schemaregistry/avro/AvroCompatibilityTest.java#L77)
- [Which just forwards the call to the `AvroSchema` constructor](https://github.com/confluentinc/schema-registry/blob/e8f89f8e55ac3f2a3c32abbd9f4de0affff238c8/core/src/main/java/io/confluent/kafka/schemaregistry/avro/AvroUtils.java#L32)
- [The `AvroSchema` will instantiate a avro `Schema.Parser`](https://github.com/confluentinc/schema-registry/blob/e8f89f8e55ac3f2a3c32abbd9f4de0affff238c8/client/src/main/java/io/confluent/kafka/schemaregistry/avro/AvroSchema.java#L119-L123), and [call use it to parser the schema](https://github.com/confluentinc/schema-registry/blob/e8f89f8e55ac3f2a3c32abbd9f4de0affff238c8/client/src/main/java/io/confluent/kafka/schemaregistry/avro/AvroSchema.java#L74)
- [Apache avro will instantiate a jackson JsonParser](https://github.com/apache/avro/blob/89c03cfe5fe8a0ad1aacf910e9b94c57a41ea67e/lang/java/avro/src/main/java/org/apache/avro/Schema.java#L1413), and [use it with a ObjectMaper](https://github.com/apache/avro/blob/89c03cfe5fe8a0ad1aacf910e9b94c57a41ea67e/lang/java/avro/src/main/java/org/apache/avro/Schema.java#L1425)
- [The ObjectMapper will deserialized the object, and only throw if the flag `FAIL_ON_TRAILING_TOKENS` is set](https://github.com/FasterXML/jackson-databind/blob/39fab67429c0f23399f031e0c0dbd4a13a73622d/src/main/java/com/fasterxml/jackson/databind/ObjectMapper.java#L2421-L2423p), which happens [to be off by default](https://github.com/FasterXML/jackson-databind/blob/39fab67429c0f23399f031e0c0dbd4a13a73622d/src/main/java/com/fasterxml/jackson/databind/DeserializationFeature.java#L238-L254)

So the above JSON is *not* invalid from the perspective of the test tooling. The next question is if it is ever invalid from the API perspective.

Reply:
It looks like this is the case, the [`AvroSchemaProvider` uses the same `AvroSchema` class to parse the json](https://github.com/confluentinc/schema-registry/blob/e8f89f8e55ac3f2a3c32abbd9f4de0affff238c8/client/src/main/java/io/confluent/kafka/schemaregistry/avro/AvroSchemaProvider.java#L53-L55)